In [1]:
import pandas as pd
import numpy as np

In [2]:
matches = pd.read_csv('/Users/tliu/Desktop/Erdos Project/Data/matches.csv')

In [3]:
matches[(matches['player1'] == 'Zhao Xintong')].tail(20)

player1              player2  score1  score2  best_of  \
107815  Zhao Xintong       Stuart Bingham       4       1        7   
108012  Zhao Xintong         Jackson Page       4       2        7   
108141  Zhao Xintong         Allan Taylor       4       2        7   
108200  Zhao Xintong        Andres Petrov       4       0        7   
108261  Zhao Xintong       Mink Nutcharut       4       2        7   
108421  Zhao Xintong              Wu Yize       5       3        9   
108485  Zhao Xintong        Mohammed Asif       5       1        9   
114104  Zhao Xintong  Akani Songsermsawad       6       3       11   
114135  Zhao Xintong            Jiang Jun       6       2       11   
114167  Zhao Xintong     Noppon Saengkham       6       2       11   
114196  Zhao Xintong         Ricky Walden       6       2       11   
115005  Zhao Xintong        Cheung Ka Wai      10       3       19   
115036  Zhao Xintong         Long Zehuang      10       8       19   
115070  Zhao Xintong          Lyu Haotian      10       4       19   
115100  Zhao Xintong       Elliot Slessor      10       8       19   
115113  Zhao Xintong            Jak Jones      10       4       19   
115129  Zhao Xintong           Peifan Lei      13      10       25   
115137  Zhao Xintong        Chris Wakelin      13       5       25   
115141  Zhao Xintong    Ronnie O'Sullivan      17       7       33   
115143  Zhao Xintong        Mark Williams      18      12       35   

        tournament_id                date  year  
107815           5163          2022-09-27  2022  
108012           5316          2022-11-28  2022  
108141           5343          2022-12-12  2022  
108200           5343          2022-12-14  2022  
108261           5344          2022-11-03  2022  
108421           5400          2022-11-21  2023  
108485           5400          2022-11-22  2023  
114104           6400          2024-11-16  2024  
114135           6400          2024-11-18  2024  
114167           6400          2024-11-19  2024  
114196           6400          2024-11-21  2024  
115005           6743  2025-04-07 - 04-08  2025  
115036           6743          2025-04-10  2025  
115070           6743          2025-04-13  2025  
115100           6743          2025-04-15  2025  
115113           6743  2025-04-20 - 04-21  2025  
115129           6743  2025-04-25 - 04-27  2025  
115137           6743  2025-04-29 - 04-30  2025  
115141           6743  2025-05-01 - 05-02  2025  
115143           6743  2025-05-04 - 05-05  2025

In [4]:
matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115630 entries, 0 to 115629
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   player1        115630 non-null  object
 1   player2        115630 non-null  object
 2   score1         115630 non-null  int64 
 3   score2         115630 non-null  int64 
 4   best_of        115630 non-null  int64 
 5   tournament_id  115630 non-null  int64 
 6   date           39002 non-null   object
 7   year           115630 non-null  int64 
dtypes: int64(5), object(3)
memory usage: 7.1+ MB


In [5]:
from elo_v2 import get_stats
# df = get_stats(matches)

In [6]:
# df

In [7]:

Ks = [5,7]
ratios = [0.05, 0.1]

for K in Ks:
    for r in ratios: 
        print(f'K = {K}, r = {r}')
        df = get_stats(matches = matches , K_factor = K, r = r)
        print(df.sort_values(by = ['elo_rating'], ascending=False).head(15))
        print(df.sort_values(by = ['elo_rating'], ascending=False).tail(10))

K = 5, r = 0.05
Before 6743, Zhao Xintong's elo is: 1710.0205226446187
After 6743, Zhao Xintong's elo is: 1938.1017405725306
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2101            1534         1106       0.0
Ronnie O'Sullivan        2036            1555         1181       0.0
Kyren Wilson             2022            1122          721       0.0
John Higgins             1986            1933         1363       0.0
Zhao Xintong             1938             375          223       0.0
Mark Selby               1908            1640         1121       0.0
Mark Williams            1896            1835         1201       0.0
Shaun Murphy             1863            1413          927       0.0
Mark Allen               1852            1125          747       0.0
Neil Robertson           1852            1331          883       0.0
Yan Bingtao              1851             354          232       0.0
Ding Junhui              1840            1033  

In [8]:
df.sort_values(by = ['elo_rating'], ascending=False).tail(20)

elo_rating  matches_played  matches_won  win_rate
David Greaves            575              56            9       0.0
Jackie Rea               569              59           11       0.0
John Dunning             568             162           45       0.0
Andrew Highfield         565              51            4       0.0
Imran Puri               563              76           10       0.0
Jim Donnelly             549             162           47       0.0
Neil Hoggarth            545              47            3       0.0
Fred Davis               545             114           27       0.0
Rogelio Esteiro          538              44            0       0.0
Maurice Parkin           513              23            0       0.0
Billy Kelly              509             109           28       0.0
Dessie Sheehan           499             184           37       0.0
Pascal Burke             481              69           15       0.0
Mike Darrington          479             132           31       0.0
Anthony Bridge           443              53            1       0.0
Dennis Hughes            441             107           21       0.0
Jack Fitzmaurice         426             179           37       0.0
Bert Demarco             389              69            6       0.0
Bernard Bennett          352              62            6       0.0
Clive Everton            312              61            4       0.0

In [9]:
df.loc['Judd Trump']

elo_rating        2163.0
matches_played    1534.0
matches_won       1106.0
win_rate             0.0
Name: Judd Trump, dtype: Float64

K = 6, r = 0.1
Before 6743, Zhao Xintong's elo is: 1762.9793645727084
After 6743, Zhao Xintong's elo is: 2034.5857183234164
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2121            1534         1106       0.0
Kyren Wilson             2057            1122          721       0.0
Ronnie O'Sullivan        2040            1555         1181       0.0
Zhao Xintong             2035             375          223       0.0
John Higgins             2035            1933         1363       0.0
Mark Williams            1950            1835         1201       0.0
Mark Selby               1948            1640         1121       0.0
Yan Bingtao              1900             354          232       0.0
Shaun Murphy             1898            1413          927       0.0
Neil Robertson           1881            1331          883       0.0
Ding Junhui              1876            1033          685       0.0
Barry Hawkins            1876            1391          870       0.0
Mark Allen               1875            1125          747       0.0
Luca Brecel              1838             676          394       0.0
Xiao Guodong             1830             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Billy Kelly              539             109           28       0.0
Pascal Burke             527              69           15       0.0
Dessie Sheehan           509             184           37       0.0
Mike Darrington          503             132           31       0.0
Anthony Bridge           495              53            1       0.0
Dennis Hughes            475             107           21       0.0
Jack Fitzmaurice         447             179           37       0.0
Bert Demarco             434              69            6       0.0
Bernard Bennett          407              62            6       0.0
Clive Everton            372              61            4       0.0
K = 8, r = 0.1
Before 6743, Zhao Xintong's elo is: 1847.0876093239517
After 6743, Zhao Xintong's elo is: 2202.145325539665
                   elo_rating  matches_played  matches_won  win_rate
Zhao Xintong             2202             375          223       0.0
Judd Trump               2201            1534         1106       0.0
John Higgins             2152            1933         1363       0.0
Kyren Wilson             2149            1122          721       0.0
Ronnie O'Sullivan        2100            1555         1181       0.0
Mark Williams            2068            1835         1201       0.0
Mark Selby               2047            1640         1121       0.0
Yan Bingtao              1986             354          232       0.0
Shaun Murphy             1985            1413          927       0.0
Neil Robertson           1966            1331          883       0.0
Ding Junhui              1964            1033          685       0.0
Barry Hawkins            1963            1391          870       0.0
Mark Allen               1943            1125          747       0.0
Xiao Guodong             1938             814          475       0.0
Luca Brecel              1933             676          394       0.0
                  elo_rating  matches_played  matches_won  win_rate
Billy Kelly              485             109           28       0.0
Maurice Parkin           462              23            0       0.0
Mike Darrington          461             132           31       0.0
Pascal Burke             440              69           15       0.0
Dennis Hughes            413             107           21       0.0
Jack Fitzmaurice         409             179           37       0.0
Anthony Bridge           396              53            1       0.0
Bert Demarco             351              69            6       0.0
Bernard Bennett          303              62            6       0.0
Clive Everton            257              61            4       0.0
K = 10, r = 0.1
Before 6743, Zhao Xintong's elo is: 1911.296029455649
After 6743, Zhao Xintong's elo is: 2348.8013548172075
                   elo_rating  matches_played  matches_won  win_rate
Zhao Xintong             2349             375          223       0.0
Judd Trump               2267            1534         1106       0.0
John Higgins             2256            1933         1363       0.0
Kyren Wilson             2218            1122          721       0.0
Mark Williams            2173            1835         1201       0.0
Ronnie O'Sullivan        2151            1555         1181       0.0
Mark Selby               2131            1640         1121       0.0
Shaun Murphy             2058            1413          927       0.0
Yan Bingtao              2050             354          232       0.0
Neil Robertson           2040            1331          883       0.0
Ding Junhui              2039            1033          685       0.0
Barry Hawkins            2037            1391          870       0.0
Xiao Guodong             2025             814          475       0.0
Luca Brecel              2016             676          394       0.0
Mark Allen               1997            1125          747       0.0
                  elo_rating  matches_played  matches_won  win_rate
Mike Darrington          435             132           31       0.0
Rogelio Esteiro          421              44            0       0.0
Jack Fitzmaurice         386             179           37       0.0
Pascal Burke             371              69           15       0.0
Dennis Hughes            370             107           21       0.0
Maurice Parkin           369              23            0       0.0
Anthony Bridge           314              53            1       0.0
Bert Demarco             290              69            6       0.0
Bernard Bennett          222              62            6       0.0
Clive Everton            164              61            4       0.0

K = 6, r = 0.05
Before 6743, Zhao Xintong's elo is: 1769.0463642217114
After 6743, Zhao Xintong's elo is: 2038.0886046345493
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2151            1534         1106       0.0
Kyren Wilson             2083            1122          721       0.0
Ronnie O'Sullivan        2074            1555         1181       0.0
John Higgins             2052            1933         1363       0.0
Zhao Xintong             2038             375          223       0.0
Mark Selby               1964            1640         1121       0.0
Mark Williams            1963            1835         1201       0.0
Shaun Murphy             1915            1413          927       0.0
Yan Bingtao              1912             354          232       0.0
Neil Robertson           1897            1331          883       0.0
Mark Allen               1896            1125          747       0.0
Ding Junhui              1893            1033          685       0.0
Barry Hawkins            1890            1391          870       0.0
Luca Brecel              1852             676          394       0.0
Xiao Guodong             1836             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Billy Kelly              536             109           28       0.0
Pascal Burke             531              69           15       0.0
Anthony Bridge           509              53            1       0.0
Dessie Sheehan           498             184           37       0.0
Mike Darrington          495             132           31       0.0
Dennis Hughes            474             107           21       0.0
Jack Fitzmaurice         437             179           37       0.0
Bert Demarco             434              69            6       0.0
Bernard Bennett          416              62            6       0.0
Clive Everton            382              61            4       0.0
K = 8, r = 0.05
Before 6743, Zhao Xintong's elo is: 1858.3091817058864
After 6743, Zhao Xintong's elo is: 2209.496474811447
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2233            1534         1106       0.0
Zhao Xintong             2209             375          223       0.0
Kyren Wilson             2180            1122          721       0.0
John Higgins             2170            1933         1363       0.0
Ronnie O'Sullivan        2136            1555         1181       0.0
Mark Williams            2083            1835         1201       0.0
Mark Selby               2063            1640         1121       0.0
Shaun Murphy             2005            1413          927       0.0
Yan Bingtao              2004             354          232       0.0
Ding Junhui              1984            1033          685       0.0
Neil Robertson           1982            1331          883       0.0
Barry Hawkins            1979            1391          870       0.0
Mark Allen               1967            1125          747       0.0
Luca Brecel              1949             676          394       0.0
Xiao Guodong             1947             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Dessie Sheehan           478             184           37       0.0
Maurice Parkin           470              23            0       0.0
Mike Darrington          447             132           31       0.0
Pascal Burke             442              69           15       0.0
Anthony Bridge           410              53            1       0.0
Dennis Hughes            408             107           21       0.0
Jack Fitzmaurice         394             179           37       0.0
Bert Demarco             346              69            6       0.0
Bernard Bennett          310              62            6       0.0
Clive Everton            265              61            4       0.0
K = 10, r = 0.05
Before 6743, Zhao Xintong's elo is: 1925.786805578041
After 6743, Zhao Xintong's elo is: 2358.544508331642
                   elo_rating  matches_played  matches_won  win_rate
Zhao Xintong             2359             375          223       0.0
Judd Trump               2301            1534         1106       0.0
John Higgins             2276            1933         1363       0.0
Kyren Wilson             2254            1122          721       0.0
Mark Williams            2190            1835         1201       0.0
Ronnie O'Sullivan        2188            1555         1181       0.0
Mark Selby               2149            1640         1121       0.0
Shaun Murphy             2080            1413          927       0.0
Yan Bingtao              2072             354          232       0.0
Ding Junhui              2061            1033          685       0.0
Neil Robertson           2059            1331          883       0.0
Barry Hawkins            2055            1391          870       0.0
Xiao Guodong             2038             814          475       0.0
Luca Brecel              2034             676          394       0.0
Mark Allen               2024            1125          747       0.0
                  elo_rating  matches_played  matches_won  win_rate
Rogelio Esteiro          431              44            0       0.0
Mike Darrington          416             132           31       0.0
Maurice Parkin           376              23            0       0.0
Pascal Burke             370              69           15       0.0
Jack Fitzmaurice         367             179           37       0.0
Dennis Hughes            361             107           21       0.0
Anthony Bridge           328              53            1       0.0
Bert Demarco             279              69            6       0.0
Bernard Bennett          225              62            6       0.0
Clive Everton            169              61            4       0.0

K = 5, r = 0.05
Before 6743, Zhao Xintong's elo is: 1710.0205226446187
After 6743, Zhao Xintong's elo is: 1938.1017405725306
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2101            1534         1106       0.0
Ronnie O'Sullivan        2036            1555         1181       0.0
Kyren Wilson             2022            1122          721       0.0
John Higgins             1986            1933         1363       0.0
Zhao Xintong             1938             375          223       0.0
Mark Selby               1908            1640         1121       0.0
Mark Williams            1896            1835         1201       0.0
Shaun Murphy             1863            1413          927       0.0
Mark Allen               1852            1125          747       0.0
Neil Robertson           1852            1331          883       0.0
Yan Bingtao              1851             354          232       0.0
Ding Junhui              1840            1033          685       0.0
Barry Hawkins            1839            1391          870       0.0
Luca Brecel              1795             676          394       0.0
Xiao Guodong             1769             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Pascal Burke             584              69           15       0.0
Billy Kelly              575             109           28       0.0
Anthony Bridge           567              53            1       0.0
Mike Darrington          529             132           31       0.0
Dennis Hughes            518             107           21       0.0
Dessie Sheehan           518             184           37       0.0
Bert Demarco             488              69            6       0.0
Bernard Bennett          480              62            6       0.0
Jack Fitzmaurice         468             179           37       0.0
Clive Everton            451              61            4       0.0
K = 5, r = 0.1
Before 6743, Zhao Xintong's elo is: 1707.5853652263318
After 6743, Zhao Xintong's elo is: 1937.4927132613764
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2073            1534         1106       0.0
Ronnie O'Sullivan        2003            1555         1181       0.0
Kyren Wilson             2000            1122          721       0.0
John Higgins             1970            1933         1363       0.0
Zhao Xintong             1937             375          223       0.0
Mark Selby               1893            1640         1121       0.0
Mark Williams            1884            1835         1201       0.0
Shaun Murphy             1847            1413          927       0.0
Yan Bingtao              1842             354          232       0.0
Neil Robertson           1835            1331          883       0.0
Mark Allen               1832            1125          747       0.0
Barry Hawkins            1825            1391          870       0.0
Ding Junhui              1825            1033          685       0.0
Luca Brecel              1783             676          394       0.0
Xiao Guodong             1765             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Pascal Burke             579              69           15       0.0
Billy Kelly              576             109           28       0.0
Anthony Bridge           554              53            1       0.0
Mike Darrington          534             132           31       0.0
Dessie Sheehan           526             184           37       0.0
Dennis Hughes            517             107           21       0.0
Bert Demarco             487              69            6       0.0
Jack Fitzmaurice         475             179           37       0.0
Bernard Bennett          470              62            6       0.0
Clive Everton            441              61            4       0.0
K = 7, r = 0.05
Before 6743, Zhao Xintong's elo is: 1817.3659178587711
After 6743, Zhao Xintong's elo is: 2127.486268016916
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2194            1534         1106       0.0
Kyren Wilson             2135            1122          721       0.0
Zhao Xintong             2127             375          223       0.0
John Higgins             2113            1933         1363       0.0
Ronnie O'Sullivan        2107            1555         1181       0.0
Mark Williams            2025            1835         1201       0.0
Mark Selby               2015            1640         1121       0.0
Shaun Murphy             1962            1413          927       0.0
Yan Bingtao              1962             354          232       0.0
Neil Robertson           1941            1331          883       0.0
Ding Junhui              1940            1033          685       0.0
Barry Hawkins            1936            1391          870       0.0
Mark Allen               1934            1125          747       0.0
Luca Brecel              1903             676          394       0.0
Xiao Guodong             1895             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Billy Kelly              504             109           28       0.0
Dessie Sheehan           485             184           37       0.0
Pascal Burke             484              69           15       0.0
Mike Darrington          468             132           31       0.0
Anthony Bridge           457              53            1       0.0
Dennis Hughes            438             107           21       0.0
Jack Fitzmaurice         413             179           37       0.0
Bert Demarco             387              69            6       0.0
Bernard Bennett          360              62            6       0.0
Clive Everton            320              61            4       0.0
K = 7, r = 0.1
Before 6743, Zhao Xintong's elo is: 1808.4236578537755
After 6743, Zhao Xintong's elo is: 2121.806749206499
                   elo_rating  matches_played  matches_won  win_rate
Judd Trump               2163            1534         1106       0.0
Zhao Xintong             2122             375          223       0.0
Kyren Wilson             2106            1122          721       0.0
John Higgins             2096            1933         1363       0.0
Ronnie O'Sullivan        2071            1555         1181       0.0
Mark Williams            2011            1835         1201       0.0
Mark Selby               2000            1640         1121       0.0
Yan Bingtao              1946             354          232       0.0
Shaun Murphy             1944            1413          927       0.0
Neil Robertson           1924            1331          883       0.0
Ding Junhui              1922            1033          685       0.0
Barry Hawkins            1921            1391          870       0.0
Mark Allen               1911            1125          747       0.0
Luca Brecel              1888             676          394       0.0
Xiao Guodong             1887             814          475       0.0
                  elo_rating  matches_played  matches_won  win_rate
Billy Kelly              509             109           28       0.0
Dessie Sheehan           499             184           37       0.0
Pascal Burke             481              69           15       0.0
Mike Darrington          479             132           31       0.0
Anthony Bridge           443              53            1       0.0
Dennis Hughes            441             107           21       0.0
Jack Fitzmaurice         426             179           37       0.0
Bert Demarco             389              69            6       0.0
Bernard Bennett          352              62            6       0.0
Clive Everton            312              61            4       0.0